src/Namespace.sol

```diff
@@ -0,0 +1,578 @@
+        }
+    }
+
+    /**
+     * @notice Returns the ETH requires to register a name for the rest of the year.
+     *
+     * @dev the fee is pro-rated for the remainder of the year by the number of seconds left.
+     */
+    function currYearFee() public returns (uint256) {
+        uint256 _currYear = currYear();
+
+        unchecked {
+            // timestampOfYear and currYear are pretermined values and cannot overflow.
+            uint256 nextYearTimestamp = timestampOfYear(_currYear + 1);
+
+            return
+                ((nextYearTimestamp - block.timestamp) * fee) /
+                (nextYearTimestamp - timestampOfYear(_currYear));
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             INTERNAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Moves the username to the destination and resets recovery state. Similar to
+     *      transferFrom but more gas-efficient since it doesn't check ownership or destination
+     *      validity which can be ensured by the caller explicitly or implicitly.
+     */
+    function _unsafeTransfer(address to, uint256 tokenId) private {
+        address from = _ownerOf[tokenId];
+
+        if (from == address(0)) revert Unauthorized();
+
+        // Underflow is prevented as long as the ownership is verified and overflow is unrealistic
+        // given the limited set of usernames available.
+        unchecked {
+            _balanceOf[from]--;
+
+            _balanceOf[to]++;
+        }
+
+        _ownerOf[tokenId] = to;
+
+        delete getApproved[tokenId];
+
+        _clearRecovery(tokenId);
+
+        emit Transfer(from, to, tokenId);
+    }
+
+    /**
+     * @dev Resets the recoveryAddress and any ongoing recoveries
+     */
+    function _clearRecovery(uint256 tokenId) private {
+        // Checking state before clearing is more gas-efficient than always clearing
+        if (recoveryClockOf[tokenId] != 0) delete recoveryClockOf[tokenId];
+
+        delete recoveryOf[tokenId];
+    }
+
+    /**
+     * @dev Returns true if the name is only composed of [a-z0-9] and the hyphen characters.
+     */
+    function _isValidUsername(bytes16 name) private pure returns (bool) {
+        uint256 length = name.length;
+
+        for (uint256 i = 0; i < length; ) {
+            uint8 charInt = uint8(name[i]);
+            // Optimize: consider using a bitmask to check for valid characters which may be more
+            // efficient.
+            // Allow inclusive ranges 45(-), 48 - 57 (0-9), 97-122 (a-z)
+            if (
+                (charInt >= 1 && charInt <= 44) ||
+                (charInt >= 46 && charInt <= 47) ||
+                (charInt >= 58 && charInt <= 96) ||
+                charInt >= 123
+            ) {
+                return false;
+            }
+
+            // i can never overflow because length is guaranteed to be < 16
```

Comment:
That check doesn't happen in this function. Might be worth adding to prevent developer errors calling this without checking

Reply:
i believe it is implicitly guaranteed -- length is calculated from .length called on a bytes16 value, so there is no way you should able to call this function and cause length to exceed 16